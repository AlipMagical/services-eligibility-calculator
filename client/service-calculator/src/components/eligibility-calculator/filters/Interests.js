/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Card, Col, Row } from 'react-bootstrap';
import React, { useContext, useState } from 'react';

import { Chip } from '../Chip';
import { DataContext } from '../../../DataContext';
import FilterModal from './FilterModal';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { getSelectedInterestNames } from '../../../filters/filterInterests';
import { useTranslation } from 'react-i18next';

function InterestCard({ interest, onClick, selected }) {
  return (
    <Col xs={6} key={interest['@id']} className="card-col">
      <Card
        onClick={onClick}
        className={selected ? 'selected' : ''}
        as="button"
        aria-pressed={selected ? 'true' : 'false'}
      >
        <Card.Body>
          <i className="material-icons card-icon align-top" aria-hidden="true">
            {interest.materialIcon || 'texture'}
          </i>
          <span className="card-text align-middle">{interest.name}</span>
        </Card.Body>
      </Card>
    </Col>
  );
}

function sendAnalytics(interests) {
  let interestList = '';
  let interestCount = 0;
  for (let value of Object.values(interests)) {
    if (interestList.length > 0) {
      interestList += ',';
    }
    interestList += value;
    interestCount++;
    ReactGA.event({
      category: 'Filter',
      action: 'Interests',
      label: value
    });
  }
  ReactGA.event({
    category: 'Filter',
    action: 'AllInterests',
    label: interestList,
    value: interestCount
  });
}

function InterestsFilterModal({
  selectedInterests,
  setSelectedInterests,
  setShow,
  interests
}) {
  const [localState, setLocalState] = useState(selectedInterests);
  const { t } = useTranslation();
  const toggleInterest = interest => {
    const value = localState[interest['@id']] ? undefined : interest.name;
    setLocalState({ ...localState, [interest['@id']]: value });
  };

  return (
    <FilterModal
      className="interests"
      show
      onApply={() => {
        setSelectedInterests(localState);
        setShow(false);
        sendAnalytics(localState);
      }}
      onHide={() => setShow(false)}
      onClear={() => {
        setShow(false);
        setSelectedInterests({});
      }}
      clearDisabled={getSelectedInterestNames(localState).length === 0}
      title={t('filterInterests.modalTitle')}
    >
      <Row className="card-row">
        {interests &&
          interests.map(interest => (
            <InterestCard
              key={interest['@id']}
              onClick={() => toggleInterest(interest)}
              selected={localState[interest['@id']]}
              interest={interest}
            />
          ))}
      </Row>
    </FilterModal>
  );
}

export default function Interests({ selectedInterests, setSelectedInterests }) {
  const [show, setShow] = useState(false);
  const { interests } = useContext(DataContext);
  const { t } = useTranslation();

  const selectedInterestNames = getSelectedInterestNames(selectedInterests);

  return (
    <>
      <Chip
        analytics={{ action: 'Interests', label: 'Add interests' }}
        active={selectedInterestNames.length > 0}
        onClick={() => setShow(true)}
      >
        {selectedInterestNames.length > 0
          ? selectedInterestNames.join(', ')
          : t('filterInterests.buttonAdd')}
      </Chip>
      {show && (
        <InterestsFilterModal
          {...{ selectedInterests, setSelectedInterests, setShow, interests }}
        />
      )}
    </>
  );
}

Interests.propTypes = {
  selectedInterests: PropTypes.object.isRequired,
  setSelectedInterests: PropTypes.func.isRequired
};
