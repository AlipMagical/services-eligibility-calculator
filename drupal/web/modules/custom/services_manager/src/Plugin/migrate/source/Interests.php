<?php
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


namespace Drupal\services_manager\Plugin\migrate\source;

use Drupal\migrate\Plugin\migrate\source\SqlBase;
use Drupal\migrate\Row;

/**
 *
 * @MigrateSource(
 *   id = "interests",
 *   source_module = "services_manager",
 * )
 */
class Interests extends SqlBase {

  /**
   * {@inheritdoc}
   */
  public function query() {
    // Query source data from the 'interest' table.
    $query = $this->select('interest', 'i')
      ->fields('i', [
          'id',
          'name',
          'material_icon',
        ]);
    return $query;
  }

  /**
   * {@inheritdoc}
   *
   * Keys are the field machine names as used in field mappings, values are
   * descriptions.
   */
  public function fields() {
    $fields = [
      'id' => $this->t('id'),
      'name' => $this->t('name'),
      'material_icon' => $this->t('material_icon'),
    ];
    return $fields;
  }

  /**
   * {@inheritdoc}
   *
   * The source fields uniquely identifying a source row.
   */
  public function getIds() {
    return [
      'id' => [
        'type' => 'uuid',
        'alias' => 'i',
      ],
    ];
  }

  /**
   * {@inheritdoc}
   */
  public function prepareRow(Row $row) {
    // convert API Platform's binary(16) id to Drupal's varchar(128) uuid
    $id = $row->getSourceProperty('id');
    $hex = bin2hex($id);
    $hex = preg_replace('/([0-9a-f]{8})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{12})/', '$1-$2-$3-$4-$5', $hex);
    $row->setSourceProperty('id', $hex);
    return parent::prepareRow($row);
  }
}
