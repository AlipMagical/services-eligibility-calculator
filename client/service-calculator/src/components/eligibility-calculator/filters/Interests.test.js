/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { act } from 'react-dom/test-utils';
import Interests from './Interests';
import { axe, toHaveNoViolations } from 'jest-axe';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { DataContext } from '../../../DataContext';

expect.extend(toHaveNoViolations);

it('renders button without selected interests', async () =>
  act(async () => {
    fetch.setupDataContextFetches();

    const { container } = render(
      <DataContext.StateProvider>
        <Interests selectedInterests={{}} setSelectedInterests={() => {}} />
      </DataContext.StateProvider>
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal without selected interests', async () =>
  act(async () => {
    fetch.setupDataContextFetches();

    const { getByText, getByRole } = render(
      <DataContext.StateProvider>
        <Interests selectedInterests={{}} setSelectedInterests={() => {}} />
      </DataContext.StateProvider>
    );

    fireEvent.click(getByText('Add interests'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));

it('renders button with a selected interest', async () =>
  act(async () => {
    fetch.setupDataContextFetches();

    const { container } = render(
      <DataContext.StateProvider>
        <Interests
          selectedInterests={{ '/api/interests/1': 'Interest1' }}
          setSelectedInterests={() => {}}
        />
      </DataContext.StateProvider>
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal with a selected interest', async () =>
  act(async () => {
    fetch.setupDataContextFetches();

    const { getByText, getByRole } = render(
      <DataContext.StateProvider>
        <Interests
          selectedInterests={{ '/api/interests/1': 'Interest1' }}
          setSelectedInterests={() => {}}
        />
      </DataContext.StateProvider>
    );

    fireEvent.click(getByText('Interest1'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));
