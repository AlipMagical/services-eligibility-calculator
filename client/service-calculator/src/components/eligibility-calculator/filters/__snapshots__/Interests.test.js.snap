// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button with a selected interest 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis active"
  >
    Interest1
  </button>
</div>
`;

exports[`renders button without selected interests 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis "
  >
    Add interests
  </button>
</div>
`;

exports[`renders modal with a selected interest 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Interests
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="card-row row"
      >
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                pets
              </i>
              <span
                class="card-text align-middle"
              >
                Animals
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="true"
            class="selected card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                child_friendly
              </i>
              <span
                class="card-text align-middle"
              >
                Childcare
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                nature_people
              </i>
              <span
                class="card-text align-middle"
              >
                Energy
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                fastfood
              </i>
              <span
                class="card-text align-middle"
              >
                Food
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                home
              </i>
              <span
                class="card-text align-middle"
              >
                Housing
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                forum
              </i>
              <span
                class="card-text align-middle"
              >
                Mediation
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                person_outline
              </i>
              <span
                class="card-text align-middle"
              >
                Older adults
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                directions_run
              </i>
              <span
                class="card-text align-middle"
              >
                Wellness
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;

exports[`renders modal without selected interests 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Interests
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="card-row row"
      >
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                pets
              </i>
              <span
                class="card-text align-middle"
              >
                Animals
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                child_friendly
              </i>
              <span
                class="card-text align-middle"
              >
                Childcare
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                nature_people
              </i>
              <span
                class="card-text align-middle"
              >
                Energy
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                fastfood
              </i>
              <span
                class="card-text align-middle"
              >
                Food
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                home
              </i>
              <span
                class="card-text align-middle"
              >
                Housing
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                forum
              </i>
              <span
                class="card-text align-middle"
              >
                Mediation
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                person_outline
              </i>
              <span
                class="card-text align-middle"
              >
                Older adults
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-6"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                directions_run
              </i>
              <span
                class="card-text align-middle"
              >
                Wellness
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="disabled btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
